### PR TITLE
OAuth Tutorials

### PR DESCRIPTION
**Pull request by ProbablePrime**
This PR aims to remove all references to username/password logon within our developer site. To this end it:

- Sets up a popup based implicit oauth flow, tested on Chrome, Edge and FireFox which replaces a template string with an implicit token.
- Removes python tutorials, (These will return when we re-write our Python clients)
- Makes all tutorials operate over OAuth.
- Removes auth from the rest tutorial - not needed
- Adds a bunch of links to the OAuth spec to the oauth reference page to encourage the oauth spec to be read
- Adds expiry times for our various token types

Video of the implict flow(older video without popup centering):
https://i.probableprime.co.uk/u/r/eqEgbvCgM5.mp4

### CODE REVIEW

File: src/js/modals.js [hunk 1]
```diff
@@ -9,6 +9,10 @@ $(document).ready(function () {
         }
 
         var anchorId = document.location.hash.substr(1); // strip #
+        //Don't do this when
```
Comment (connor4312) [inline src/js/modals.js]: ?

File: src/js/oauth.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    // Puts focus on the newWindow
+    if (window.focus) {
+        newWindow.focus();
+    }
+}
+
+function openImplicitOAuthWindow (scopes) {
+    var clientId = 'fa54866255ea641235e596e5659fa726a4aa9f7ecc72758f';
+    var redirectURI = 'https://dev.beam.pro/oauthreturn.html';
+    // var redirectURI = 'http://localhost:3000/oauthreturn.html';
```
Comment (connor4312) [inline src/js/oauth.js]: 🔥

File: src/js/oauth.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
+'use strict';
```
Comment (connor4312) [inline src/js/oauth.js]: Please wrap this in an IIFE to avoid global leaks

File: src/tutorials/code/java/chatbot/6.java [hunk 4]
```diff
@@ -15,11 +15,9 @@
 
 public class Chat {
     public static void main(String[] args) throws ExecutionException, InterruptedException {
-        String username = args[0];
-        String password = args[1];
-        BeamAPI beam = new BeamAPI();
+        BeamAPI beam = new BeamAPI("AUTH_TOKEN");
 
-        BeamUser user = beam.use(UsersService.class).login(username, password).get();
+        BeamUser user = beam.use(UsersService.class)..getCurrent().get();
```
Comment (connor4312) [inline src/tutorials/code/java/chatbot/6.java]: `..`

File: src/tutorials/interactive.pug [hunk 5]
```diff
@@ -120,35 +124,6 @@ block tutorialContent
...
-            p.
-                After that, we'll write our error and report handlers. Our report handler
-                interacts with the mouse based on Interactive inputs.
-            +highlightFile('python','./tutorials/code/python/interactive/3.py')
-            p.
-                The final step is to setup the coroutine that will run everything.
-            +highlightFile('python','./tutorials/code/python/interactive/4.py')
-            p.
-                The full code is shown below.
-            +highlightFile('python','./tutorials/code/python/interactive/5.py')
```
Comment (connor4312) [inline src/tutorials/interactive.pug]: Why kill the Python interactive tutorial? This is still valid and the python client is used.

File: src/js/oauth.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    if (window.focus) {
+        newWindow.focus();
+    }
+}
+
+function openImplicitOAuthWindow (scopes) {
+    var clientId = 'fa54866255ea641235e596e5659fa726a4aa9f7ecc72758f';
+    var redirectURI = 'https://dev.beam.pro/oauthreturn.html';
+    // var redirectURI = 'http://localhost:3000/oauthreturn.html';
+    var url = 'https://beam.pro/oauth/authorize?response_type=token&' +
```
Comment (SimonSchick) [inline src/js/oauth.js]: You can probably clean this up using the `URL` object (doesn't work in IE 11 tho).

File: src/tutorials/interactive.pug (hunk 5 above)
Comment (ProbablePrime) [inline src/tutorials/interactive.pug]: I left a request for comments in #dev-ecosystem and had no response. The interactive client doesnt work. It cannot send progress updates.

Comment (connor4312) [inline src/tutorials/interactive.pug]: We should probably fix that then instead of just removing it! 😄

File: src/js/oauth.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+ */
+function getAccessToken () {
+    var hash = window.location.hash.slice(1);
+
+    if (hash.length === 0) {
+        return null;
+    }
+
+    var hashParts = hash.split('&');
+    for (let i = 0; i < hashParts.length; i++) {
```
Comment (SimonSchick) [inline src/js/oauth.js]: You can use `hashParts.find`

File: src/tutorials/interactive.pug (hunk 5 above)
Comment (ProbablePrime) [inline src/tutorials/interactive.pug]: We discussed that and there are no resources to do so. Removing it seemed the better option for now.

File: src/tutorials/code/java/chatbot/6.java (hunk 4 above)
Comment (SimonSchick) [inline src/tutorials/code/java/chatbot/6.java]: syntax error.

File: src/tutorials/interactive.pug (hunk 5 above)
Comment (connor4312) [inline src/tutorials/interactive.pug]: bleh. Ah well. Interactive 2!

File: build/misc.js [hunk 8]
```diff
@@ -67,7 +67,8 @@ function getLocals () {
         bsTabs: {},
         highlight: (lang, str) => {
             if (lang === 'text') return str;
-            return highlight.highlight(lang, str).value;
+            return highlight.highlight(lang, str).value
+                .replace('AUTH_TOKEN', '<a class="auth-token">Click here to get your Token!</a>');
```
Comment (ProbablePrime) [inline build/misc.js]: This passes lint :S

File: src/js/oauth.js (hunk 7 above)
Comment (ProbablePrime) [inline src/js/oauth.js]: Id rather not without polyfills for `.find` in place

Comment (SimonSchick) [inline src/js/oauth.js]: https://kangax.github.io/compat-table/es6/ Do we support IE11 for the docs site?

Comment (ProbablePrime) [inline src/js/oauth.js]: thats a no:


Adding .find and URL object

Comment (SimonSchick) [inline src/js/oauth.js]: Disregard this and the URL object, seems like we are stuck with IE11 for now.
I required GA for the site so maybe we can get some insights.

Comment (ProbablePrime) [inline src/js/oauth.js]: Yup i agree, lets get some tracking on the issue before we change our supported browser matrix

**Comment by ProbablePrime** [comment]
Alright I think everything is covered, please review again

**Comment by ProbablePrime** [comment]
I added https://trello.com/c/WzutVMed/92-fix-ie11-for-the-developer-site to track the IE11 situtation

**Comment by ProbablePrime** [comment]
@connor4312 this is ready for a another look from you. Know you're busy but would love to see it out there :)

File: src/js/oauth.js [hunk 9]
```diff
@@ -1,90 +1,92 @@
...
-    window.close();
-}
+    function handleOAuthResponse () {
+        var token = getAccessToken();
 
-function replaceToken (token) {
-    if (!token) {
-        $('.auth-token').html('Error retrieving token');
-        return;
+        window.opener.replaceToken(token);
```
Comment (connor4312) [inline src/js/oauth.js]: Does this still work? replaceToken probably needs to be explicitly attached to the window now

Comment (ProbablePrime) [inline src/js/oauth.js]: Oh... oh... good catch i was sleepy :)

**Comment by ProbablePrime** [comment]
@connor4312 now it works :)

**Pull request merged**